File: westcoast/src/main/java/frc/commands/Autonomous.java
package frc.commands;


import edu.wpi.first.math.util.Units;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.Command;
import frc.commands.drive.LeaveLine;
import frc.commands.magazine.Shoot;
import frc.commands.targeting.AlignToTarget;
import frc.commands.targeting.SetupShot;
import frc.robot.definition.Definition;
import frc.vision.LimelightRunner;
import frc.vision.VisionRunner;
import org.jetbrains.annotations.NotNull;

import java.util.function.BooleanSupplier;

import static frc.commands.Commands.IntakeCommands.collect;
import static frc.commands.Commands.IntakeCommands.deploy;
import static frc.commands.Commands.ShooterCommands.shoot;

public final class Autonomous {
    @NotNull
    public static Command move(@NotNull Definition definition, double distance) {
        return new LeaveLine(
                distance,
                definition.controllers.leaveLine(),
                definition.controllers.leaveLine(),
                definition.subsystems.driveTrain
        );
    }

    @NotNull
    public static Command setupShot(@NotNull Definition definiton, double distance){
        return new SetupShot(
                definiton,
                distance
        );
    }

    public static Command shootAndLeave(@NotNull Definition definition, double speed) {
        return shoot(definition, () -> speed, () -> Shoot.Mode.Full)
                .andThen(move(definition, 1.5));
    }

    public static Command visionAuton(
            @NotNull Definition definition,
            @NotNull BooleanSupplier ballAcceptanceStrategy,
            double speed) {
        return deploy(definition)
                .andThen(
                        move(definition, 1.0)
                                .alongWith(
                                        collect(definition, ballAcceptanceStrategy)
                                )
                                .withTimeout(2.5)
                                .until(definition.subsystems.magazine::hasIndexSensorBreak)
                )
                .andThen(setupShot(definition, Units.inchesToMeters(55)))
                    .withTimeout(3)
                .andThen(shoot(definition, () -> speed, () -> Shoot.Mode.Full));
                //.andThen(move(definition, 1.5));
    }

    public static Command shootOutPickInShootOut(
            @NotNull Definition definition,
            @NotNull BooleanSupplier ballAcceptanceStrategy,
            double speed
    ) {
        return shoot(definition, () -> speed, () -> Shoot.Mode.Full)
                .alongWith(deploy(definition))
                .andThen(
                        move(definition, 1.0)
                                .alongWith(collect(definition, ballAcceptanceStrategy))
                                .until(definition.subsystems.magazine::hasFeederSensorBreak)
                )
                .andThen(move(definition, -1.0))
                .andThen(shoot(definition, () -> speed, () -> Shoot.Mode.Full))
                //shoot never stops so might have to fix that
                .andThen(move(definition, 1.5));
    }

    public static Command outPickInShootTwice(
            @NotNull Definition definition,
            @NotNull BooleanSupplier ballAcceptanceStrategy,
            double speed
    ) {
        return deploy(definition)
                .andThen(
                        move(definition, 1.0)
                                .alongWith(
                                        collect(definition, ballAcceptanceStrategy)
                                )
                                .withTimeout(2.5)
                                .until(definition.subsystems.magazine::hasIndexSensorBreak)
                )
                .andThen(move(definition, -1.0))
                .andThen(shoot(definition, () -> speed, () -> Shoot.Mode.Full))
                .andThen(move(definition, 1.5));
    }

    public static Command shootLeaveAndCollect(
            @NotNull Definition definition,
            @NotNull BooleanSupplier ballAcceptanceStrategy,
            double speed
    ) {
        return shoot(definition, () -> speed, () -> Shoot.Mode.Full)
                .alongWith(deploy(definition))
                .andThen(
                        move(definition, 1.5)
                                .alongWith(collect(definition, ballAcceptanceStrategy))
                                .until(definition.subsystems.magazine::hasFeederSensorBreak)
                );
    }
}

File: westcoast/src/main/kotlin/frc/commands/Autonomous.kt
package frc.commands

import edu.wpi.first.wpilibj2.command.Command
import frc.commands.Commands.IntakeCommands.collect
import frc.commands.Commands.IntakeCommands.deploy
import frc.commands.Commands.ShooterCommands.shoot
import frc.commands.drive.LeaveLine
import frc.commands.magazine.Shoot
import frc.robot.definition.Definition

object Autonomous {
    fun Definition.move(distance: Double): Command {
        return LeaveLine(
            distance,
            controllers.leaveLine(),
            controllers.leaveLine(),
            subsystems.driveTrain
        )
    }

    fun Definition.shootAndLeave(speed: Double): Command {
        return shoot(
            velocity = { speed },
            mode = { Shoot.Mode.Full }
        ).andThen(move(1.5))
    }

    fun Definition.shootOutPickInShootOut(
        ballAcceptanceStrategy: () -> Boolean,
        speed: Double
    ): Command {
        return shoot(
            velocity = { speed },
            mode = { Shoot.Mode.Full }
        ).alongWith(deploy())
            .andThen(
                move(1.0)
                    .alongWith(collect(ballAcceptanceStrategy))
                    .until { subsystems.magazine.hasFeederSensorBreak() }
            )
            .andThen(move(-1.0))
            .andThen(shoot({ speed }, { Shoot.Mode.Full }))
            .andThen(move(1.5))
    }

    fun Definition.outPickInShootTwice(
        ballAcceptanceStrategy: () -> Boolean,
        speed: Double
    ): Command {
        return deploy()
            .andThen(
                move(1.0)
                    .alongWith(collect(ballAcceptanceStrategy))
                    .withTimeout(2.5)
                    .until { subsystems.magazine.hasIndexSensorBreak() }
            )
            .andThen(move(-1.0))
            .andThen(shoot({ speed }, { Shoot.Mode.Full }))
            .andThen(move(1.5))
    }

    fun Definition.shootLeaveAndCollect(
        ballAcceptanceStrategy: () -> Boolean,
        speed: Double
    ): Command {
        return shoot({ speed }, { Shoot.Mode.Full })
            .alongWith(deploy())
            .andThen(
                move(1.5)
                    .alongWith(collect(ballAcceptanceStrategy))
                    .until { subsystems.magazine.hasFeederSensorBreak() }
            )
    }
}